libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -869,8 +869,10 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 
 	if (functionCallKind == FunctionCallKind::TypeConversion)
 	{
+		Type::Category category = _functionCall.expression().annotation().type->category();
 		solAssert(
-			_functionCall.expression().annotation().type->category() == Type::Category::TypeType,
+			category == Type::Category::TypeType ||
+			category == Type::Category::UserDefinedValueType,
```

Comment:
I don't think this special case is needed if the type checker properly assigns a typetype to `MyType` in the expression `MyType(t)`